### ISSUE TITLE
Design a way to automatically load algorithm implementations

### ISSUE THREAD

**Issue by stevenroose** [PointyCastle/pointycastle#50]
_From @izaera on January 26, 2014 18:33_

The idea is to avoid the need to call initCipher() before using the library. I'll try to call initCipher() from Registry.create().

Also, it would be good to redesign initCipher() so that it can be tweaked to load only the algorithms used. This is a good starting point: http://blog.sethladd.com/2013/04/lazy-load-libraries-in-dart.html

_Copied from original issue: izaera/cipher#50_

**Comment by stevenroose** [comment]
_From @izaera on January 27, 2014 9:29_

According to this: http://feedproxy.google.com/~r/sethladd/aGPE/~3/okOCiSHeifE/minification-is-not-enough-you-need.html  algorithms which are not used (nor registered with cipher's registry) wouldn't be sent through the wire when in client side.

**Comment by stevenroose** [comment]
So this is basically what I tried to tell you in the previous e-mail? Ok,
nice :)

Btw, the method of mailing a third mail as a response to the first e-mail
is a much more efficient way of making me read the second e-mail first than
adding it to the subject line and eventually still sending that third
e-mail :D I was genuinly confused because the READ THIS ONE FIRST was
actually the first (bottom) one in the list.

(I'm ranting again, did a good exam and rewarded me a glass of whisky. I'll
try to be more concise and formal again like I usually am to strangers :p)

**Comment by stevenroose** [comment]
_From @izaera on January 28, 2014 19:57_

I tried to provide an specialized "impl" library which would export all algorithms so that the use can cherry pick which one to uses and bypass the registry. But I have reverted it. The main reason is that currently I have three impl.dart files: the base one, the client and the server. Thus, I have to make three of those specialized libraries. And even in that case there are many deployment combinations I leave out so I just decided that I would provide the whole lib or the alternative to do it by cherrypicking and optionally registering the implementations with the registry if the user wants to use it. Having so many libraries is error prone and I'm sure I will forget to add the implementations to each one of them.

The idea is that the user of the library can tailor a customized dart file in his project so that he only needs to import that file in the rest of his libraries.

I've also moved impl.dart, impl_server.dart and impl_client.dart to a separate folder to make the library cleaner.
